src/app/payments/get-protocol-fee.ts

```diff
@@ -0,0 +1,142 @@
+import { decodeInvoice, defaultTimeToExpiryInSeconds } from "@domain/bitcoin/lightning"
+import { checkedToWalletId } from "@domain/wallets"
+import {
+  LnPaymentRequestNonZeroAmountRequiredError,
+  LnPaymentRequestZeroAmountRequiredError,
+  PriceRatio,
+} from "@domain/payments"
+import { LndService } from "@services/lnd"
+import { PaymentFlowStateRepository } from "@services/payment-flow"
+import { WalletsRepository } from "@services/mongoose"
+import { NewDealerPriceService } from "@services/dealer-price"
+import { addAttributesToCurrentSpan } from "@services/tracing"
+
+import {
+  constructPaymentFlowBuilder,
+  newCheckIntraledgerLimits,
+  newCheckWithdrawalLimits,
+} from "./helpers"
+
+export const getLightningFeeEstimation = async ({
+  walletId,
+  paymentRequest,
+}: {
+  walletId: string
+  paymentRequest: EncodedPaymentRequest
+}): Promise<PaymentAmount<WalletCurrency> | ApplicationError> => {
+  const decodedInvoice = decodeInvoice(paymentRequest)
+  if (decodedInvoice instanceof Error) return decodedInvoice
+  if (decodedInvoice.paymentAmount === null) {
+    return new LnPaymentRequestNonZeroAmountRequiredError()
+  }
+
+  return estimateLightningFee({
+    uncheckedSenderWalletId: walletId,
+    invoice: decodedInvoice,
+  })
+}
+
+export const getNoAmountLightningFeeEstimation = async ({
+  walletId,
+  paymentRequest,
+  amount,
+}: {
+  walletId: string
+  paymentRequest: EncodedPaymentRequest
+  amount: number
+}): Promise<PaymentAmount<WalletCurrency> | ApplicationError> => {
+  const decodedInvoice = decodeInvoice(paymentRequest)
+  if (decodedInvoice instanceof Error) return decodedInvoice
+
+  const { amount: lnInvoiceAmount } = decodedInvoice
+  if (lnInvoiceAmount && lnInvoiceAmount > 0) {
+    return new LnPaymentRequestZeroAmountRequiredError()
+  }
+
+  return estimateLightningFee({
+    uncheckedSenderWalletId: walletId,
+    invoice: decodedInvoice,
+    uncheckedAmount: amount,
+  })
+}
+
+const estimateLightningFee = async ({
+  uncheckedSenderWalletId,
+  invoice,
+  uncheckedAmount,
+}: {
+  uncheckedSenderWalletId: string
+  invoice: LnInvoice
+  uncheckedAmount?: number
+}): Promise<PaymentAmount<WalletCurrency> | ApplicationError> => {
+  const senderWalletId = checkedToWalletId(uncheckedSenderWalletId)
+  if (senderWalletId instanceof Error) return senderWalletId
+
+  const senderWallet = await WalletsRepository().findById(senderWalletId)
+  if (senderWallet instanceof Error) return senderWallet
+
+  const dealer = NewDealerPriceService()
+  const builder = await constructPaymentFlowBuilder({
+    senderWallet,
+    invoice,
+    uncheckedAmount,
+    usdFromBtc: dealer.getCentsFromSatsForImmediateBuy,
```

Comment:
Shouldn't this be the option price?

Reply:
Done: https://github.com/GaloyMoney/galoy/pull/1237/commits/ac2b75038b876ec0e86ea892faf5aee680997d1b